### PR TITLE
api: fix pickling of custom sparse functions with more than one dim

### PR DESCRIPTION
**Pull request by mloubout**
Fix pickling of sparse function that don't use shape_global as rkwargs which makes a strong assumption there is a single dimension:

https://github.com/devitocodes/devito/blob/19d61feced13aff2f5bcc260b69565e66bb5f6af/devito/types/sparse.py#L102

### CODE REVIEW

File: devito/types/sparse.py [hunk 1]
```diff
@@ -95,12 +95,14 @@ def __shape_setup__(cls, **kwargs):
...
         if grid is None:
             raise TypeError('Need `grid` argument')
-        shape = kwargs.get('shape')
+        shape = kwargs.get('shape', kwargs.get('shape_global'))
         dimensions = kwargs.get('dimensions')
         npoint = kwargs.get('npoint', kwargs.get('npoint_global'))
         glb_npoint = SparseDistributor.decompose(npoint, grid.distributor)
         if shape is None:
             loc_shape = (glb_npoint[grid.distributor.myrank],)
+        elif dimensions is None:
```
Comment (FabioLuporini) [inline devito/types/sparse.py:107]: I don't understand how this is possible and, even then, what it would mean to assign `loc_shape = shape`. If it's an obscure case could you add a short comment?

File: devito/types/sparse.py [hunk 2]
```diff
@@ -742,7 +745,11 @@ def __shape_setup__(cls, **kwargs):
                 raise ValueError('`nt` must be > 0')
 
             shape = list(AbstractSparseFunction.__shape_setup__(**kwargs))
-            shape.insert(cls._time_position, nt)
+            try:
```
Comment (FabioLuporini) [inline devito/types/sparse.py]: using assert for control flow is unpythonic (asserts can be stripped with -O and should signal programmer errors, not drive logic).

File: devito/types/sparse.py [hunk 3]
```diff
@@ -742,7 +745,11 @@ def __shape_setup__(cls, **kwargs):
...
+                # Rebuilt with full shape
+                assert shape[cls._time_position] == nt
+            except AssertionError:
+                shape.insert(cls._time_position, nt)
```
Comment (FabioLuporini) [inline devito/types/sparse.py]: Im also not sure I understand when can we end up here? perhaps it's simply because I don't understand this bug yet... and this is entirely on me

File: devito/types/sparse.py (hunk 2 above)
Comment (FabioLuporini) [reply devito/types/sparse.py]: also, in the unlucky circumstance that `nt == x_size (+ halo etc)` we would have an issue here no?

I think we need to check `len(shape)` instead...

Or even better, ensure we don't need to worry about this at this point, which seems a bit weird

File: devito/types/sparse.py (hunk 1 above)
Comment (mloubout) [reply devito/types/sparse.py:107]: This is only possible when rebuilding without dimension, in which case the shape is already the correct one

File: devito/types/sparse.py (hunk 3 above)
Comment (mloubout) [reply devito/types/sparse.py]: We end up here when we create a SparseTimeFunction which uses `nt` to infer the time size. At rebuild it has both nt and shape. Can just ignore shape for time function but that'd still need a try/except

File: devito/types/sparse.py (hunk 1 above)
Comment (FabioLuporini) [reply devito/types/sparse.py:107]: but why would one rebuild w/o dimensions?

Comment (mloubout) [reply devito/types/sparse.py:107]: People don't it's just some very deep corner case we have in a test to make sure rebuild works correctly.

**Review by FabioLuporini — APPROVED** [review]
